### ISSUE TITLE
Nested component isn't detected by the Page->hasComponent method.

### ISSUE THREAD

**Issue by RomainMazB** [octobercms/october#5406]
- OctoberCMS Build:471
- PHP Version:7.4

### Description:
When calling a component inside of antoher component, the page can't correctly detect the nested component.

### Steps To Reproduce:
I just tried to play around with Mall plugin which uses this techniques in all its components.
Example in its checkout component:
![image](https://user-images.githubusercontent.com/53976837/102254289-baba6c80-3f08-11eb-9631-96bb455fdcd7.png)

So It can easily be reproduced installing it and trying to call this on the checkout page:
```
Event::listen('cms.page.initComponents', function ($controller, $page, $layout) use ($componentManager) {
    dump($page->components);
    dump($page->hasComponent('shippingMethodSelector'));
    dump($page->hasComponent('OFFLINE\Mall\Components\ShippingMethodSelector'));
});
```
the result:
![image](https://user-images.githubusercontent.com/53976837/102253962-51d2f480-3f08-11eb-8e82-53f34495a6ff.png)

This is probably due to the CompoundObject context and may be related to an (very) old one issue: [#2619](https://github.com/octobercms/october/issues/2619)

**Comment by mjauvin** [comment]
Looks like the PR that fixed issue #2619 only did this when rendering a component from the page markup, but did not address a partial that contains a component.

This would be in Cms/Classes/Controller::renderPartial() here: https://github.com/octobercms/october/blob/develop/modules/cms/classes/Controller.php#L997-L998

The previousContext should be saved here as well.

**Comment by mjauvin** [comment]
Or maybe here would make more sense:
https://github.com/octobercms/october/blob/develop/modules/cms/classes/ComponentBase.php#L140-L146

Can you test this and submit a PR @RomainMazB ?

**Comment by mjauvin** [comment]
Maybe the controller's `setComponentContext()` should save the previous context and a new `restoreComponentContext()` should use the saved context and restore it?

**Comment by mjauvin** [comment]
Should we have a queue to store these to support multi depth nesting?

**Comment by RomainMazB** [comment]
Yep, I'll test it in few minutes 👍🏼 .

I didn't take the time to dive into the previous issue/PR for now, wanted to report before the end of my day session ;)

**Comment by RomainMazB** [comment]
After private investigations with @mjauvin :
the problem comes from the fact the `hasComponent` method is only [looping over the `$this->settings['components']` array](https://github.com/octobercms/october/blob/develop/modules/cms/classes/CmsCompoundObject.php#L235) to say if a component is present or not.

[`$this->settings['components']` is initialized with only the components present into the page ini block](https://github.com/octobercms/october/blob/develop/modules/cms/classes/CmsCompoundObject.php#L184-L203).

Due to that, the `hasComponent` has no recursivity method and can't detect any component called by other component and directly [rendered by the twig compiler](https://github.com/octobercms/october/blob/09b7e1b2fd75b46284c9517fe2503f4574b17638/modules/cms/twig/Extension.php#L122-L125).

Obviously, it would be a huge lack of performance to look into each component's rendered partial to detect and add components recursively.

Plus a work-around could be easily done using the the keys of `$page->components` array which is correctly filled with all the rendered components.

I let @LukeTowers decide to close the issue or suggest any different thought on that.

**Comment by daftspunk** [comment]
The referenced change (#2619) just allowed other components to call upon other components within the markup. All components must be registered in either the layout or the page to ensure they are fully registered. Partial components are restricted by their [life cycle limitations](http://octoberdev.test/docs/cms/partials#life-cycle-limitations), as you correctly put, it would be a significant resource expense to try to proactively scan for these since they are only registered at runtime

The idea is that
1. Component A is added to page/layout
2. Compo B is registered in Compo A's PHP code (now available to `hasComponent`)
3. Compo A now needs to be able to use Compo B (this is the reason we merged #2619)

Just consider this to be a design limitation and move the component registration into the pages or layouts that use them, instead of inside the partials. Hope this makes sense, if this needs more attention just ping me and I'll reopen
